src/CachePeers.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_CACHEPEERS_H
+#define SQUID_SRC_CACHEPEERS_H
+
+#include "CachePeer.h"
+#include "mem/PoolingAllocator.h"
+
+#include <memory>
+#include <vector>
+
+/// cache_peer configuration storage
+class CachePeers
+{
+public:
+    /// owns stored CachePeer objects
+    using Storage = std::vector< std::unique_ptr<CachePeer>, PoolingAllocator< std::unique_ptr<CachePeer> > >;
+
+    /// stores a being-configured cache_peer
+    void add(CachePeer *p) { storage.emplace_back(p); }
+
+    /// deletes a previously add()ed CachePeer object
+    void remove(CachePeer *);
+
+    /// the number of currently stored (i.e. added and not removed) cache_peers
+    size_t size() const { return storage.size(); }
+
+    /* peer iterators forming a sequence for C++ range-based for loop API */
+    using const_iterator = Storage::const_iterator;
+    const_iterator begin() const { return storage.cbegin(); }
+    const_iterator end() const { return storage.cend(); }
+
+    /// A CachePeer to query next when scanning all peer caches in hope to fetch
+    /// a remote cache hit. Never nil. \sa neighborsUdpPing()
+    /// \param iteration a 0-based index of a loop scanning all peers
+    CachePeer *nextPeerToPing(size_t iteration);
+
+private:
+    /// cache_peers in configuration/parsing order
+    Storage storage;
+
+    /// total number of completed peer scans by nextPeerToPing()-calling code
+    uint64_t peerPolls_ = 0;
+};
+
+/// All configured cache_peers that are still available/relevant.
+/// \returns an empty container if no cache_peers were configured or all
+/// configured cache_peers were removed (e.g., by DeleteConfigured()).
+const CachePeers &CurrentCachePeers();
+
+/// destroys the given peer after removing it from the set of configured peers
+void DeleteConfigured(CachePeer *);
```

Comment:
Calls to this should be replaced by `CurrentCachePeers()->erase(Foo);`

Reply:
In summary, AFAICT, the current arguments in support of this change request are:

* The additional code duplication that the requested changes will create will be small (for now).
* The existing callers will not assert (because they are working correctly).
* The existing callers do not care about the invariant protected by the assertion.
 
There is no consensus regarding the validity of the last argument listed above.

The arguments in support of PR code include:

* The change request will lead to compiler errors. Fixing those compiler errors while providing safe constant peer iteration is not trivial, likely to lead to more change requests.
* PR code reduces code duplication.
* PR code helps alert developers and reviewers to the associated code risks.
* The assertions may detect new bugs in future callers. Official code has bugs in the same area.

Please resolve this change request without further arguments because the existing arguments, after two review iterations, are still not even close to justifying the requested changes.

A detailed response follows.

----

> > > Amos: Calls to this should be replaced by `CurrentCachePeers()->erase(Foo);`
 
> > Alex: That replacement is not possible because CurrentCachePeers() provides safe read-only access.

@yadij, please note that this change request is asking for a minor change (i.e. removing a simple wrapper function) that is not technically _possible_ (without additional work to make the changed code compile while preserving safety guarantees). Even if this change is implemented despite the lack of consensus, this change request creates non-trivial work for very little (and disputed) gain. Please consider resolving this change request without further arguments on those grounds alone.

> > However, even if we were to remove or work around that safety feature, why should those calls be replaced? I believe the current solution is significantly better:

> > * The caller does not just want to erase an item from a collection. The caller wants to destroy the CachePeer object as well. The difference can be critical and providing visual clues about it (in the form of the word "delete") is a good idea IMO.
 
> Amos: The STL API distinction between `remove` and `erase` is already well-known. We can safely abide by that without needing our home-grown wrappers and terminology.

The alleged well-known difference between STL remove() and STL erase() or our alleged ability to do "abide by that difference" is not the point here! The point is that we want the caller to know that they are deleting (including in C++ sense) the cache peer object. The proposed refactoring will not let the caller know that, even if the caller is perfectly aware of well-known facts (which _is_ a reasonable design assumption).

> > * The caller wants to delete (what the caller believes to be) a configured peer P. To avoid code that deletes wrong peers, we should assert that P is one of the configured peers. In STL containers, erase() variations that accept objects tolerate objects that are not present in the container -- those erase() methods return the number of removed elements (0 or 1). We do not want to imply that this removal has similar forgiving semantics.

> AFAICT none of our code cares (nor needs to) about whether the `CachePeer` object is part of the list or not before the action. All that matters is that it is not there afterward.

While a caller might indeed not care[^1] or, much more likely, may forget to care, this wrapper defends a useful invariant -- those who remove cache peers from the current configuration (a major, disruptive action!) specify a peer that actually exists in the current configuration. A buggy caller may leave the peer object they do want to remove while specifying some stale peer pointer. This wrapper will detect such problems. If we start hitting these assertions, we may eventually decide to relax the guarantees, but it is best to _start_ with reasonable strict guarantees that are more likely to catch bugs. Future relaxing is easier than future restricting!

[^1]: Whether existing callers do not care is debatable, but I am trying to avoid a yet another debate. 

> > * Unlike the proposed replacement, DeleteConfigured() does not result in caller code duplication.
 
> There are only two callers in current code.

Yes, the suggested code duplication will be (currently) small, but it is still duplication. Even small duplication often leads to bugs. I should not be defending the lack of duplication! It is a clear positive.


> Amos: * `neighborsUdpAck()` has a raw-pointer being used before and after the act of deletion. If that peer is not listed we have bigger issues, and do not care since it is being dropped from availability effective immediately. Adding the assert/Assure looks okay, but IMO is paranoid.

The assertion will detect some of those bigger issues if they appear. I care about that. The assertion may be a bit paranoid for a given caller fixed by this PR, but it is there for all callers, including future ones. It is not paranoid when all callers are considered. It protects a useful invariant. As discussed above, we may relax that invariant later if needed, but that possibility does not make this assertion paranoid (quite the opposite, actually!).

> * `neightbors_init()` is used by startup/reconfigure. All objects are guaranteed to be part of the list,

Reconfiguration is a long process. By the time neightbors_init() is called, some peers might have been already removed or invalidated. However, those potential complications are not actually relevant here! neightbors_init() iterates currently configured cache_peers. Naturally, it will never trigger the assertion designed to detect pointers to non-currently-configured cache_peers unless its iteration itself becomes very buggy (e.g., if it tries to delete the same peer twice). In that case, the assertion would be very useful!